src/main/ruby/truffleruby/core/posix.rb

```diff
@@ -425,11 +425,52 @@ def self.read_string_native(io, length)
     end
   end
 
-  def self.read_to_buffer_native(io, length)
+  def self.execute_posix_read(io, buffer, length)
     fd = io.fileno
+    return Truffle::POSIX.read(fd, buffer, length) unless io.timeout
+
+    timeout_milliseconds = Truffle::KernelOperations.convert_duration_to_milliseconds(io.timeout)
+    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond) + timeout_milliseconds
+
+    while true
+      bytes_read = Truffle::POSIX.read(fd, buffer, length)
+      return bytes_read if bytes_read >= 0
+      return bytes_read unless Errno.errno == Errno::EAGAIN::Errno
+
+      current_timeout = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
+      raise IO::TimeoutError if current_timeout <= 0
+
+      poll_result = Truffle::IOOperations.poll_milliseconds(io, Truffle::IOOperations::POLLIN, current_timeout)
+      raise IO::TimeoutError if poll_result == 0
+      Errno.handle_errno(Errno.errno) if poll_result == -1
+    end
```

Comment:
I wonder if we should move the timeout logic in another method, to keep `execute_posix_read` small for the common case of no timeout. It would matter if there is splitting going on for instance. But there is probably no need to split this.
Still it feels a bit cleaner to move the logic to e.g. `execute_posix_read_with_timeout(io, buffer, length)`, could you do that and also for write?

Reply:
Sure - though I'm only guessing what you meant by this as it's hard to imagine how splitting would work in the middle of a Ruby side method. Please check and let me know if I did it right.

Reply:
Yes this looks great, it notably means lower memory footprint if IO#timeout are never used (probably common since this is a fairly new API with rather specific use cases)